### PR TITLE
feat: Add a /graphql and /graphiql endpoints that exposes an integrations list.

### PR DESCRIPTION
**Pull request by chirino**

### CODE REVIEW

**Comment by chirino** [comment]
@zregvart @deeleman @rhuss here's an initial spike of what it would look like to add a GraphQL interface in front of our normal APIs.

Using this we basically define the view model using GraphQL schema and then the UI defines which fields from the objects they are interested in when doing requests via their queries.

This is here basically just to explore the idea.  WDYT?

File: app/rest/runtime/src/main/java/io/syndesis/runtime/GraphQLConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+public class GraphQLConfiguration {
+
+    @Autowired
+    IntegrationHandler integrationHandler;
+
+    @Bean
+    GraphQLSchema schema() throws IOException {
+
+        String schema;
+        try (InputStream is = Thread.currentThread().getContextClassLoader().getResourceAsStream("schema.graphqls") )  {
```
Comment (iocanel) [inline app/rest/runtime/src/main/java/io/syndesis/runtime/GraphQLConfiguration.java]: Why do we rely on TCCL when we can just do 

GraphQLConfiguration.class.getClassLoader().getResourceAsStream("shema.graphqls")?

**Comment by chirino** [comment]
Works both ways.

**Comment by iocanel** [comment]
Sure, but isn't it a little bit safer if we didn't rely on TCCL? Frameworks screw with TCCL all the time.

**Comment by chirino** [comment]
We know exactly what framework we are using.  Don't think it's an issue.

**Comment by deeleman** [comment]
I see pros and cons on using GraphQL. Basically it gives the UI a god tool to query the application data graph and get taylored responses that match the UI needs more closely. However its use is justified if the UI wants to perform queries that encompass different related entities and get an unified response model schema. This becomes  specially useful only if the UI is aware of the application data graph tree and is willing to conduct advanced queries amongst different related entities.

If we only plan to use GraphQL to submit Http request to fetch data which is already formalized as an existing model in the backend, and leverage GraphQL to filter out unwanted properties, then we do not need since we're already doing that by means of the client models.

So question here is: Will we want the UI to be aware of the backend application data graph in the future, so it can query the API as if it was querying a database? If so (that would probably save a lot of time to the backend folks by delegating that responsibility on the UI devs), then this is the way to go.

All in all, exploring this idea and implementing this quick POC would be a great way to see its real value. The only caveat is that leveraging GraphQL will require us to implement the ApolloQL library on the frontend and configure it, and we'll need to allocate some time for it.

File: app/rest/runtime/src/main/java/io/syndesis/runtime/GraphQLConfiguration.java (hunk 1 above)
Comment (kahboom) [reply app/rest/runtime/src/main/java/io/syndesis/runtime/GraphQLConfiguration.java]: Have to agree with @iocanel here, why risk the headache if we don't need to, and it's easy to change?

**Comment by jimmidyson** [comment]
Just a quick pedantic aside:

> Add a /graphql and /graphiql rest endpoints

GraphQL != REST

**Comment by kahboom** [comment]
Can someone help me understand this a bit better? I can't provide an educated opinion considering I'm not educated on GraphQL and ApolloQL. I was aware of it and what it intends to solve, but judging based on what I'm seeing, it just doesn't seem like something we need. It appears that it could replace our implementation of Web Sockets, and give the UI added flexibility, but it's a big change that seems wouldn't be easy to undo. I'm interested in hearing more about the benefits.

I think this doesn't resolve it in a way I thought it would. What we need is true decoupling (#540 😂 ), and us working with the DB directly (not directly, but querying the API almost as if we were) is the complete opposite of decoupling 😂-- it's just passing through one abstraction layer as if it didn't exist, putting more responsibility on the UI.

I'm all for added responsibility in the UI if it solves an existing problem. We need something that will allow a consistent interface (can our new models currently do that @deeleman ?), an available and common data structure we can work with if the REST API is borked or still undergoing changes we need to continue work in the UI.

Maybe I misunderstood something though, and am happy to be enlightened. :)

**Comment by chirino** [comment]
@deeleman @kahboom so gaphql solves the issue of getting the UI exactly the right amount of data that it need to render the current page without needing to do multiple http requests. 

It also helps formalize the shape of the data models passed back and forth between UI and backend since you can define this with the schemea. 

Keep in mind that queries are just a simple http request, you don’t need extra libs to use it.  Just peek at the test that this PR has in it. 

I think we will always have full rest resource apis, and the UI can choose which access method is better for the job. 

An interesting thing this POC highlights is that we have the ability to transform the DB model to a slightly different shape if needed. For example you can’t use objects as map data container in graphql so we convert configuredProperties to an array.

**Comment by chirino** [comment]
@jimmidyson true.. true

**Comment by rhuss** [comment]
My feelings are mixed as opposed to a decoupled approach with dedicated view objects and a translation layer in between.

Advantages of GraphQL:

* Flexible querying on an existing domain object, giving the UI more freedom to pick only what is required.
* No translation in between, so no extra layer.

Disadvantages: 

* Couples directly to the domain model, so every change in the domain model needs to be communicated and potentially leads to changes in the UI code. That is not necessarily the case with dedicated, use case specific, view objects.
* Potential performance issues as iiur all objects for a query are loaded into memory and then filtered 
(iiur the current implementation right)
* More burden put onto the UI as the ui team needs to know in detail the domain model.
* Does 'API by contract' can work with this as its just a generic API on the domain models? Is a 'contract' which can be verified possible that way?
* It's late in the release cycle, and we, unfortunately, don't have the luxury to do much experimentation and exploration (looking at all the other POCs we need to do).
* Nobody of us has any real word experience with GraphQL (maybe wrong here, though)

What would be the alternatives to improve the API situation? @zregvart you mentioned that contract based, 'pure' REST API development would be an option, too ? Is this something we can introduce incrementally ?

as said at the beginning, i have mixed feelings, so not sure whether we should go down this road. However IMO its mostly the @syndesisio/ui-api which should drive the API decision.

**Comment by chirino** [comment]
@rhuss I disagree it "Couples directly to the domain model".  The view models are defined in the [schema.graphqls](https://github.com/chirino/syndesis/blob/8f68464241072446ea95795033d4e5bf5dff0f97/app/rest/runtime/src/main/resources/schema.graphqls#L30).  Notice that while similar, it's not the exact model we use in the DB.  I probably should have made an even more drastic modification of the model to prove the decoupling point.  There is a translation layer happening, but it's mostly automatic if the UI model is matching the DB model.

The UI folks could be the ones that define the models in the schema.graphqls, and the back end folks figure out how to map to it via per field data fetchers in the [GraphQLConfiguration.java](https://github.com/chirino/syndesis/blob/8f68464241072446ea95795033d4e5bf5dff0f97/app/rest/runtime/src/main/java/io/syndesis/runtime/GraphQLConfiguration.java#L66) file.  Naturally, if parts of the UI module and the DB model match up, then there is less transformation work needed to fulfill the query request.

It's up to the data fetchers to figure out how much to fetch into memory.  They could become more DB aware to filter down what's loaded from the DB.  This is no different from our current rest APIs.  Note that the query operations at the top level can also take parameters to control things like paging.  So I don't think that's  an issue.

**Comment by rhuss** [comment]
@zregvart @chirino @deeleman Is it ok to close this PR for the moment (or park it on a branch in the upstream repo ?) I think we are heading to another direction (swagger based), so I wonder whether we should put more efforts here for the momentn.

**Comment by chirino** [comment]
fine with me.

**Pull request closed without merging**